web/blueprints/infrastructure/__init__.py

```diff
@@ -123,20 +142,261 @@ def switch_show_json(switch_id):
+                      level=switch.host.room.level, room_number=switch.host.room.number)
+
+    if form.validate_on_submit():
+        room = Room.q.filter_by(number=form.room_number.data,
+                                level=form.level.data, building=form.building.data).one()
+
+        edit_switch(switch, form.name.data, form.management_ip.data, room, current_user)
+
+        session.session.commit()
+
+        flash("Die Switch wurde erfolgreich bearbeitet.", "success")
+
+        return redirect(url_for('.switches'))
+
+    form_args = {
+        'form': form,
+        'cancel_to': url_for('.switches')
+    }
+
+    return render_template('generic_form.html',
+                           page_title="Switch bearbeiten",
+                           form_args=form_args)
+
+
+@bp.route('/switch/<int:switch_id>/delete', methods=['GET', 'POST'])
+@access.require('infrastructure_change')
+def switch_delete(switch_id):
+    switch = Switch.q.filter_by(host_id=switch_id).one()
+
+    if not switch:
+        flash(u"Switch mit ID {} nicht gefunden!".format(switch_id), "error")
+        return redirect(url_for('.switches'))
+
+    form = Form()
+
+    if form.validate_on_submit():
+        delete_switch(switch, current_user)
+
+        session.session.commit()
+
+        flash("Die Switch wurde erfolgreich gelöscht.", "success")
+
+        return redirect(url_for('.switches'))
+
+    form_args = {
+        'form': form,
+        'cancel_to': url_for('.switches'),
+        'submit_text': 'Löschen',
+        'actions_offset': 0
+    }
+
+    return render_template('generic_form.html',
+                           page_title="Switch löschen",
+                           form_args=form_args)
+
+
+@bp.route('/switch/<int:switch_id>/port/create', methods=['GET', 'POST'])
+@access.require('infrastructure_change')
+def switch_port_create(switch_id):
+    switch = Switch.q.get(switch_id)
+
+    if not switch:
+        flash(u"Switch mit ID {} nicht gefunden!".format(switch_id), "error")
+        return redirect(url_for('.switches'))
+
+    form = SwitchPortForm()
+
+    form.patch_port.query = PatchPort.q.filter_by(switch_room=switch.host.room)
+
+    if form.validate_on_submit():
+        error = False
+
+        switch_port = create_switch_port(switch, form.name.data, current_user)
+
+        if form.patch_port.data:
+            try:
+                patch_switch_port_to_patch_port(switch_port, form.patch_port.data, current_user)
+            except PatchPortAlreadyPatchedException:
+                form.patch_port.errors.append("Dieser Patch-Port ist bereits mit einem Switch-Port verbunden.")
+                error = True
+
+        if not error:
+            session.session.commit()
+
+            flash("Der Switch-Port {} wurde erfolgreich erstellt.".format(switch_port.name), "success")
+
+            return redirect(url_for('.switch_show', switch_id=switch.host_id))
+        else:
+            session.session.rollback()
```

Comment:
@ibot3 FYI this could be done simpler using the [else-clause](https://docs.python.org/3/reference/compound_stmts.html#try) of a `try..except` block:
```python
try:
    if form.patch_port.data:
        patch_switch_port_to_patch_port(switch_port, form.patch_port.data,
                                        processor=current_user)
except PatchPortAlreadyPatchedException:
    form.patch_port.errors.append("Dieser Patch-Port ist bereits mit einem Switch-Port verbunden.")
    session.session.rollback()
else:
    session.session.commit()
    flash(f"Der Switch-Port {switch_port.name} wurde erfolgreich erstellt.", 'success')
    return redirect(url_for('.switch_show', switch_id=switch.host_id))
```